### PR TITLE
Close scaffolder stale task executions

### PR DESCRIPTION
**Pull request by OscarDHdz**
Signed-off-by: OscarDHdz <v-ohernandez@expediagroup.com>
Solves: https://github.com/backstage/backstage/issues/7370

## Hey, I just made a Pull Request!

Shutdown stale scaffolder task executions:
* The "issue-scenario" happens when a new deployment is running while a `task` is being executed. The `task` is _killed_ and looses its closure. Basically this PR tries to solve that by properly closing the task when it is requested
* Currently, this is what users see (screenshot):

> A task that never ends, which make users believe that something is still running. 
image

**Changes**
* Fixed `DatabaseTaskStore.listStaleTasks` PG query
* Updated `DatabaseTaskStore` to include `shutdownTask` method which:
   * Closes any hung `processing` step (`task_events` table) - So that UI does show the hung step as Failed
   * Closes the task itself (`tasks` table)  
* Added `backend-task` under `createRouter` to close stale tasks on a schedule basis 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
As much as I quite like that we're doing this on a request basis, I wonder if it just makes sense for us to add something that uses the `backend-tasks` and just a cleanup job that runs and checks for jobs older than x rather than building this into the interface? Thoughts @backstage/maintainers?

**Comment by Rugvip** [comment]
Yep agreed. I think the check for stale tasks is a very fast DB operation too so it might not need help from `backend-tasks`, but then again if the clean isn't atomic then that'd be the simplest way around that. We're not necessarily talking a super quick detection of stale tasks either right? Should be enough to do a check every couple of minutes or so.

**Comment by OscarDHdz** [comment]
Sounds good, 
Just to double check...

* This  "check/action" should happen in a `backend-task` that runs every X minutes, instead of happening in the `GET /task` request. 

Right?

**Comment by Rugvip** [comment]
Yep! and by "backend task" we mean some something scheduled through https://github.com/backstage/backstage/tree/master/packages/backend-tasks

**Comment by OscarDHdz** [comment]
Changes are ready to be reviewed 😄 
* Moved logic from `getTask` to a `BackendTask`
* Made a small fix to the `listStaleTasks` method for the PG Query. It was complaining that `dateadd` was not a function

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts [hunk 1]
```diff
@@ -262,6 +264,20 @@ export class StorageTaskBroker implements TaskBroker {
     );
   }
 
+  /**
+   * {@inheritdoc TaskBroker.closeStaleTasks}
+   */
+  async closeStaleTasks(): Promise<void> {
+    const timeoutS =
+      this.config.getOptionalNumber('scaffolder.taskTimeout.seconds') || 300;
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:272]: Tbh it might make sense to just use some sensible default for this and skipping the possibility to configure it? I'm thinking 1h or so is well over any reasonable task duration but it'll still clean things up pretty quick.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Thanks! 👍 

One critical piece here is that we don't make any breaking changes to the API. Wherever we can it's best to just keep things internal instead. I believe it's just the `TaskStore` where that might be tricky. Couple of other comments on configurability as well, feeling that it's best just have some sensible settings and skip the added noise.

File: plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts (hunk 1 above)
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/scaffolder/tasks/StorageTaskBroker.ts:272]: Makes sense, I was expecting some feedback on that. 
I'll remove that configuration 😉

**Comment by OscarDHdz** [comment]
Post Changes summary:

* REMOVED `taskTimeout` from `app-config`
* UPDATED  `DatabaseTaskStore` to make `shutdownTask` OPTIONAL
* MOVED out `closeStaleTasks ` from `TaskBroker` and  placed it into the `createRouter` function

File: plugins/scaffolder-backend/api-report.md [hunk 2]
```diff
@@ -610,6 +615,8 @@ export interface TaskBroker {
   // (undocumented)
   claim(): Promise<TaskContext>;
   // (undocumented)
+  closeStaleTasks(): Promise<void>;
```
[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder-backend/api-report.md]: Had to make this little tweak in the `createRouter`: 
```javascript
let databaseTaskStore: DatabaseTaskStore;
if (!options.databaseTaskStore) {
  databaseTaskStore = await DatabaseTaskStore.create({ database });
} else {
  databaseTaskStore = options.databaseTaskStore;
}
```

so that later I could execute the scheduler like: `if (databaseTaskStore.shutdownTask) { ... }`

The options I had to move the logic to the `router` 🤔 :
* Instantiate once more `databaseTaskStore` in here to pass it to the `backend-task`
* Change `taskBroker` so that I could extract the `databaseTaskStore` from it
* Passing `databaseTaskStore` also within the `createRouter` options - Which was what I did on the last commit

File: plugins/scaffolder-backend/src/service/router.ts [hunk 3]
```diff
@@ -67,6 +68,7 @@ export interface RouterOptions {
   actions?: TemplateAction<any>[];
   taskWorkers?: number;
   taskBroker?: TaskBroker;
+  databaseTaskStore?: DatabaseTaskStore;
```
[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/service/router.ts:71]: I thought about it, and actually we fall in that scenario 😆 
We are using our own `taskBroker`, so would be missing this feature

On the other side, I'll change its name

File: plugins/scaffolder-backend/src/service/router.ts [hunk 4]
```diff
@@ -203,6 +211,25 @@ export async function createRouter(
   actionsToRegister.forEach(action => actionRegistry.register(action));
   workers.forEach(worker => worker.start());
 
+  if (databaseTaskStore.shutdownTask) {
+    const scheduler = TaskScheduler.fromConfig(config).forPlugin('scaffolder');
```
[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/service/router.ts:215]: Makes sense, I'll make that update

**Comment by OscarDHdz** [comment]
Changes:
* Using `scheduler` from `createRouter` param instead of creating a new one
* Renamed `DatabaseTaskBroker` to `taskBroker`
* Log cleanup 😆

File: plugins/scaffolder-backend/src/service/router.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/service/router.ts]: I thought about this during the weekend and agreed with you on running it only when no `taskBroker` is provided (default).

I was overthinking to make it "available" for everybody, but if someone is using a custom `taskBoroker` then I guess they could do something similar on their side.

**Comment by OscarDHdz** [comment]
Changes after last commit:
* Closing task now only runs if no `taskBroker` is provided (default)

**Review by Rugvip — APPROVED** [review]
Nice, thank you! 👍 

Think it's good to ship, just a clarifying question and want to let @benjdlambert have a look too

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts [hunk 5]
```diff
@@ -268,8 +269,7 @@ export class DatabaseTaskStore implements TaskStore {
         '<=',
         this.db.client.config.client.includes('sqlite3')
           ? this.db.raw(`datetime('now', ?)`, [`-${timeoutS} seconds`])
-          : this.db.raw(`dateadd('second', ?, ?)`, [
-              `-${timeoutS}`,
+          : this.db.raw(`? - interval '${timeoutS} seconds'`, [
```
[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:272]: PostgressSQL DB was throwing that `dateadd` is not a function 😕 
so I had to use the `+/- interval` approach to add time to a date

[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:272]: Just to chime in a little bit here, would like to dig in a little bit as to why `dateadd` doesn't work for you. Is it something to do with the PostgreSQL version? this works for most people right now, so it's unfortunate that it's not working for you. I mean the replacement passes the tests too, so it's probably fine, but I need some closure :joy:

[the remark this answers is not in the document]
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:272]: Sure, 
so here's what I get when I try to run the function with a PG Client:
image

Same issue was being thrown when running `listStaleTasks`. 

I am using `latest` PG container, which is `v14.5`  according to their [Docker hub page](https://hub.docker.com/_/postgres).

On the other side, I couldn't find a `dateadd` operator in PG docs for "functions-datetime": 
* https://www.postgresql.org/docs/current/functions-datetime.html

Something that I noticed about `listStaleTasks` is that it was only being called by `TaskBroker.vacuumTasks` which is not being used acros Backstage code 🤔

**Comment by OscarDHdz** [comment]
Fixed merge conflicts 😎

File: plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (Rugvip) [reply plugins/scaffolder-backend/src/scaffolder/tasks/DatabaseTaskStore.ts:272]: tbh I think it might be that the existing code is broken. Afaik postgres doesn't have a `dateadd` function in any version. It seems like the only DB tests that cover this code only runs for SQLite. Listing stale tasks isn't something that's ever done until now, so it's entirely possible that it simply doesn't work atm.

tl;dr: :shipit:

**Review by Rugvip — APPROVED** [review]
🎉, let's :shipit:

File: plugins/scaffolder-backend/src/service/router.ts [hunk 6]
```diff
@@ -166,11 +169,29 @@ export async function createRouter(
...
     taskBroker = new StorageTaskBroker(databaseTaskStore, logger);
+
+    if (scheduler && databaseTaskStore.listStaleTasks) {
+      await scheduler.scheduleTask({
+        id: 'close_stale_tasks',
+        frequency: { cron: '*/5 * * * *' }, // every 5 minutes, also supports Duration
+        timeout: { minutes: 15 },
+        fn: async () => {
+          const { tasks } = await databaseTaskStore.listStaleTasks({
+            timeoutS: 3600,
```
Comment (Rugvip) [inline plugins/scaffolder-backend/src/service/router.ts:185]: Realizing that there's a non-0 change that someone might hit this limit tbh. Perhaps it's best to bump it up to something much safer, like 24h? Especially since the current implementation won't actually cancel ongoing tasks, meaning that the impact of shutting down an active task can be quite high due to the confusion.

**Review by Rugvip — CHANGES_REQUESTED** [review]
Just one tiny tweak we realized we wanted to do after some discussion. Then I'd say we're ready to :shipit: after the release tomorrow.

File: plugins/scaffolder-backend/src/service/router.ts (hunk 6 above)
Comment (OscarDHdz) [reply plugins/scaffolder-backend/src/service/router.ts]: Makes sense, I've pushed the bump to `86400`

**Review by Rugvip — APPROVED** [review]
Nice! Time to :shipit: 😁 

Couple of merge conflicts crept in that need to be dealt with though :<

**Comment by OscarDHdz** [comment]
Fixed merge conflicts 😄
